File: Cpp_src/medium/maximum_product_of_word_lengths_318.cc
#include <iostream>
#include <vector>
using namespace std;

// Use bit manipulation
// Time Complexity: O(n^2)
// Space Complexity:O(n)
// Runtime: 48 ms, faster than 76.35%
// Memory Usage: 13 MB, less than 80.00%
class Solution
{
public:
    int maxProduct(vector<string> &words)
    {
        int result = 0;
        vector<int> value(words.size(), 0);
        for (int i = 0; i < words.size(); i++)
        {
            for (char c : words[i])
            {
                value[i] |= (1 << (c - 'a'));
            }
        }
        for (int i = 0; i < words.size(); i++)
        {
            for (int j = i + 1; j < words.size(); j++)
            {
                if ((value[i] & value[j]) == 0)
                {
                    result = max(result, int(words[i].length() * words[j].length()));
                }
            }
        }
        return result;
    }
};

File: Golang_src/medium/maximumProductOfWordLengths_318.go
package medium

// Use Bit Manipulation:
// Time Complexity: O(n^2)
// Space Complexity:O(n)
// Runtime: 12 ms, faster than 91.89%
// Memory Usage: 6.1 MB, less than 13.51%
func maxProduct(words []string) int {
	result := 0
	size := len(words)
	tmp := make([]int, size)

	for i, word := range words {
		str := []rune(word)

		for _, ch := range str {
			tmp[i] |= 1 << (ch - 'a')
		}
	}

	for i := 0; i < size; i++ {
		for j := i + 1; j < size; j++ {
			if tmp[i]&tmp[j] == 0 {
				result = max(result, len(words[i])*len(words[j]))
			}
		}
	}

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
